webapp/src/components/common/DataGridForm.tsx

```diff
@@ -0,0 +1,357 @@
+/**
+ * Copyright (c) 2024, RTE (https://www.rte-france.com)
+ *
+ * See AUTHORS.txt
+ *
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ *
+ * SPDX-License-Identifier: MPL-2.0
+ *
+ * This file is part of the Antares project.
+ */
+
+import type { IdType } from "@/common/types";
+import {
+  GridCellKind,
+  type Item,
+  type DataEditorProps,
+  type GridColumn,
+  type FillHandleDirection,
+} from "@glideapps/glide-data-grid";
+import type { DeepPartial } from "react-hook-form";
+import { FormEvent, useCallback, useEffect, useMemo, useState } from "react";
+import DataGrid, { DataGridProps } from "./DataGrid";
+import {
+  Box,
+  Divider,
+  IconButton,
+  setRef,
+  SxProps,
+  Theme,
+  Tooltip,
+} from "@mui/material";
+import useUndo, { type Actions } from "use-undo";
+import UndoIcon from "@mui/icons-material/Undo";
+import RedoIcon from "@mui/icons-material/Redo";
+import SaveIcon from "@mui/icons-material/Save";
+import { useTranslation } from "react-i18next";
+import { LoadingButton } from "@mui/lab";
+import { mergeSxProp } from "@/utils/muiUtils";
+import * as R from "ramda";
+import { SubmitHandlerPlus } from "./Form/types";
+import useEnqueueErrorSnackbar from "@/hooks/useEnqueueErrorSnackbar";
+import useFormCloseProtection from "@/hooks/useCloseFormSecurity";
+import { useUpdateEffect } from "react-use";
+
+type GridFieldValuesByRow = Record<
+  IdType,
+  Record<string, string | boolean | number>
+>;
+
+export interface DataGridFormState {
+  isDirty: boolean;
+  isSubmitting: boolean;
+}
+
+export interface DataGridFormApi<TFieldValues extends GridFieldValuesByRow> {
+  data: TFieldValues;
+  setData: Actions<TFieldValues>["set"];
+  formState: DataGridFormState;
+}
+
+export interface DataGridFormProps<
+  TFieldValues extends GridFieldValuesByRow = GridFieldValuesByRow,
+  SubmitReturnValue = unknown,
+> {
+  defaultData: TFieldValues;
+  columns: ReadonlyArray<GridColumn & { id: keyof TFieldValues[number] }>;
+  rowMarkers?: DataGridProps["rowMarkers"];
+  allowedFillDirections?: FillHandleDirection;
+  enableColumnResize?: boolean;
+  onSubmit?: (
+    data: SubmitHandlerPlus<TFieldValues>,
+    event?: React.BaseSyntheticEvent,
+  ) => void | Promise<SubmitReturnValue>;
+  onSubmitSuccessful?: (
+    data: SubmitHandlerPlus<TFieldValues>,
+    submitResult: SubmitReturnValue,
+  ) => void;
```

Comment:
```ts
 onSubmitSuccessful?: (
    data: SubmitHandlerPlus<TData>,
    submitResult?: TData,
  ) => void;
```

Reply:
No `submitResult` can be anything. It's based on the API return value.

Reply:
hum seems weird to me, "result" imply its the returned value of the submit handler, which is suppose to return data or nothing. but I will check it more in depth